File: LeetCode/1-100/34. 在排序数组中查找元素的第一个和最后一个位置.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
  let flagStart = true
  let resArr = [-1, -1]
  for (let i = 0; i < nums.length; ++i) {
    if (flagStart && nums[i] === target) {
      flagStart = false
      resArr[0] = i
    }
    if (!flagStart && nums[i] === target) {
      resArr[1] = i
    }
  }
  return resArr
}

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange2 = function (nums, target) {
  const binarySearch = (nums, target, lower) => {
    let left = 0,
      right = nums.length - 1,
      answer = nums.length
    while (left <= right) {
      const middle = Math.floor((left + right) / 2)
      if (nums[middle] > target || (lower && nums[middle] >= target)) {
        right = middle - 1
        answer = middle
      } else {
        left = middle + 1
      }
    }
    return answer
  }
  let result = [-1, -1]
  const leftIndex = binarySearch(nums, target, true)
  const rightIndex = binarySearch(nums, target, false) - 1
  if (nums[leftIndex] === target && nums[rightIndex] === target) {
    result = [leftIndex, rightIndex]
  }
  return result
}

File: LeetCode/1-100/34. 在排序数组中查找元素的第一个和最后一个位置.java
class Solution {
   public int[] searchRange(int[] nums, int target) {
        int left = binarySearch(nums,target,true);
        int right = binarySearch(nums,target,false) - 1;
        int[] ans = new int[]{-1,-1};
        if (nums.length == 0) return ans;
        if (left <= right && nums[left] == target && nums[right] == target) {
            ans[0]  =left;
            ans[1] = right;
        }
        return ans;
    }

    public int binarySearch(int[] nums,int target, boolean flagL) {
        int l = 0,r = nums.length - 1,an = nums.length;
        while (l <= r) {
            int mid = l +(r - l)/2;
            if (nums[mid] > target || (flagL && nums[mid] >= target)) {
                r = mid - 1;
                an = mid;
            } else {
                l = mid + 1;
            }
        }
        return an;
    }
}